File: jarvis.py
import ollama
import sys
from datetime import datetime

try:
    import pyttsx3
    TTS_AVAILABLE = True
except ImportError as e:
    TTS_AVAILABLE = False
    print(f"Warning: pyttsx3 not installed. Error: {e}")
    print(f"Run this command to install for your current Python:")
    print(f"  {sys.executable} -m pip install pyttsx3")
    print()

try:
    import speech_recognition as sr
    STT_AVAILABLE = True
except ImportError as e:
    STT_AVAILABLE = False
    print(f"Warning: speech_recognition not installed. Error: {e}")
    print(f"Run this command to install for your current Python:")
    print(f"  {sys.executable} -m pip install SpeechRecognition pyaudio")
    print()

import threading

def print_separator():
    print("\n" + "-" * 60 + "\n")

def print_assistant_message(message):
    print(f"\033[94mJarvis:\033[0m {message}")

def print_user_message(message):
    print(f"\033[92mSir:\033[0m {message}")

def print_error(message):
    print(f"\033[91mError:\033[0m {message}")

def print_info(message):
    print(f"\033[93m{message}\033[0m")

class TTSEngine:
    def __init__(self):
        if not TTS_AVAILABLE:
            raise Exception("pyttsx3 not available")
        
        self.rate = 175
        self.volume = 0.9
    
    def speak(self, text):
        """Speak text in a separate thread to avoid blocking"""
        def _speak():
            try:
                engine = pyttsx3.init()
                engine.setProperty('rate', self.rate)
                engine.setProperty('volume', self.volume)
                engine.say(text)
                engine.runAndWait()
                engine.stop()
            except Exception as e:
                print_error(f"TTS error: {str(e)}")
        
        thread = threading.Thread(target=_speak)
        thread.daemon = True
        thread.start()
    
    def speak_blocking(self, text):
        """Speak text and wait for completion"""
        try:
            engine = pyttsx3.init()
            engine.setProperty('rate', self.rate)
            engine.setProperty('volume', self.volume)
            engine.say(text)
            engine.runAndWait()
            engine.stop()
        except Exception as e:
            print_error(f"TTS error: {str(e)}")

def main():
    model = "llama2-uncensored:7b" 
    conversation_history = []
    tts_enabled = True
    
    tts = None
    if TTS_AVAILABLE:
        try:
            tts = TTSEngine()
            print_info("Text-to-Speech initialized successfully!")
        except Exception as e:
            print_error(f"Could not initialize TTS: {str(e)}")
            print_info("Continuing without voice output...")
            tts_enabled = False
    else:
        print_info("TTS not available. Install with: pip install pyttsx3")
        tts_enabled = False
    
    conversation_history.append({
        'role': 'system',
        'content': 'You are Jarvis, a helpful AI assistant. You are speaking with your user, whom you address as "Sir". Please share your thoughts on any topic you find interesting however stay on track and answer the users questions directly.'
    })
    
    print("\n" + "=" * 60)
    print("Jarvis v0.8")
    print("=" * 60 + "\n")
    
    if tts and tts_enabled:
        tts.speak("Jarvis online. How may I assist you, Sir?")
    
    try:
        ollama.list()
    except Exception as e:
        print_error("Cannot connect to Ollama. Make sure Ollama is running.")
        print_info("Start Ollama with: ollama serve")
        sys.exit(1)
    
    while True:
        try:
            user_input = input("\033[92mSir:\033[0m ").strip()
            
            if user_input.lower() in ['exit', 'quit']:
                print_info("Goodbye!")
                if tts and tts_enabled:
                    tts.speak_blocking("Goodbye, Sir.")
                break
            
            if user_input.lower() == 'clear':
                conversation_history = [conversation_history[0]] 
                print_info("Conversation history cleared!")
                if tts and tts_enabled:
                    tts.speak("Conversation history cleared.")
                continue
            
            if user_input.lower() == 'save':
                timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
                filename = f"conversation_{timestamp}.txt"
                with open(filename, 'w') as f:
                    for msg in conversation_history:
                        if msg['role'] != 'system': 
                            f.write(f"{msg['role'].upper()}: {msg['content']}\n\n")
                print_info(f"Conversation saved to {filename}")
                if tts and tts_enabled:
                    tts.speak(f"Conversation saved to {filename}")
                continue
            
            if user_input.lower() == 'mute':
                tts_enabled = False
                print_info("Voice output muted.")
                continue
            
            if user_input.lower() == 'unmute':
                if tts:
                    tts_enabled = True
                    print_info("Voice output enabled.")
                    tts.speak("Voice output enabled.")
                else:
                    print_info("TTS not available.")
                continue
            
            if not user_input:
                continue
            
            conversation_history.append({
                'role': 'user',
                'content': user_input
            })
            
            print("\033[94mJarvis:\033[0m ", end="", flush=True)
            
            response_content = ""
            stream = ollama.chat(
                model=model,
                messages=conversation_history,
                stream=True
            )
            
            for chunk in stream:
                content = chunk['message']['content']
                print(content, end='', flush=True)
                response_content += content
            
            print()  
            
            conversation_history.append({
                'role': 'freindly assistant',
                'content': response_content
            })
            
            if tts and tts_enabled:
                tts.speak(response_content)
            
            print_separator()
            
        except KeyboardInterrupt:
            print("\n")
            print_info("Interrupted. Type 'exit' to quit or continue chatting.")
            continue
        except Exception as e:
            print_error(f"An error occurred: {str(e)}")
            import traceback
            traceback.print_exc()
            continue

if __name__ == "__main__":
    try:
        main()
    except Exception as e:
        print_error(f"Fatal error: {str(e)}")
        import traceback
        traceback.print_exc()
    finally:
        print("\n" + "=" * 60)
        print("Press Enter to close...")
        print("=" * 60)

        input()

File: jarvis.go
package main

import (
	"bufio"
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"os/exec"
	"strings"
	"time"
)

const (
	ColorReset  = "\033[0m"
	ColorRed    = "\033[91m"
	ColorGreen  = "\033[92m"
	ColorYellow = "\033[93m"
	ColorBlue   = "\033[94m"
)

type Message struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

type ChatRequest struct {
	Model    string    `json:"model"`
	Messages []Message `json:"messages"`
	Stream   bool      `json:"stream"`
}

type ChatResponse struct {
	Message struct {
		Content string `json:"content"`
	} `json:"message"`
	Done bool `json:"done"`
}

type OllamaClient struct {
	baseURL string
	client  *http.Client
}

func NewOllamaClient(baseURL string) *OllamaClient {
	return &OllamaClient{
		baseURL: baseURL,
		client:  &http.Client{Timeout: 300 * time.Second},
	}
}

func (o *OllamaClient) Chat(model string, messages []Message) (string, error) {
	reqBody := ChatRequest{
		Model:    model,
		Messages: messages,
		Stream:   true,
	}

	jsonData, err := json.Marshal(reqBody)
	if err != nil {
		return "", err
	}

	resp, err := o.client.Post(o.baseURL+"/api/chat", "application/json", bytes.NewBuffer(jsonData))
	if err != nil {
		return "", fmt.Errorf("cannot connect to Ollama. Make sure Ollama is running")
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("Ollama returned status: %d", resp.StatusCode)
	}

	var fullResponse strings.Builder
	decoder := json.NewDecoder(resp.Body)

	for {
		var chatResp ChatResponse
		if err := decoder.Decode(&chatResp); err == io.EOF {
			break
		} else if err != nil {
			return "", err
		}

		fmt.Print(chatResp.Message.Content)
		fullResponse.WriteString(chatResp.Message.Content)

		if chatResp.Done {
			break
		}
	}

	fmt.Println()
	return fullResponse.String(), nil
}

func (o *OllamaClient) TestConnection() error {
	resp, err := o.client.Get(o.baseURL + "/api/tags")
	if err != nil {
		return fmt.Errorf("cannot connect to Ollama. Make sure Ollama is running with: ollama serve")
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("Ollama returned status: %d", resp.StatusCode)
	}

	return nil
}

type VoiceEngine struct {
	ttsEnabled bool
}

func NewVoiceEngine() *VoiceEngine {
	return &VoiceEngine{
		ttsEnabled: true,
	}
}

func (v *VoiceEngine) Speak(text string) {
	if !v.ttsEnabled {
		return
	}

	go func() {
		psScript := fmt.Sprintf(`Add-Type -AssemblyName System.Speech; $synth = New-Object System.Speech.Synthesis.SpeechSynthesizer; $synth.Rate = 1; $synth.Volume = 90; $synth.Speak('%s')`, 
			strings.ReplaceAll(text, "'", "''"))
		
		cmd := exec.Command("powershell", "-Command", psScript)
		cmd.Run()
	}()
}

func (v *VoiceEngine) SpeakBlocking(text string) {
	if !v.ttsEnabled {
		return
	}

	psScript := fmt.Sprintf(`Add-Type -AssemblyName System.Speech; $synth = New-Object System.Speech.Synthesis.SpeechSynthesizer; $synth.Rate = 1; $synth.Volume = 90; $synth.Speak('%s')`, 
		strings.ReplaceAll(text, "'", "''"))
	
	cmd := exec.Command("powershell", "-Command", psScript)
	cmd.Run()
}

func (v *VoiceEngine) Listen() (string, error) {
	printInfo("Listening... (speak now)")
	
	psScript := `
Add-Type -AssemblyName System.Speech
$recognizer = New-Object System.Speech.Recognition.SpeechRecognitionEngine
$recognizer.SetInputToDefaultAudioDevice()
$grammar = New-Object System.Speech.Recognition.DictationGrammar
$recognizer.LoadGrammar($grammar)
$result = $recognizer.Recognize()
if ($result -ne $null) {
    $result.Text
} else {
    ""
}
$recognizer.Dispose()
`
	
	cmd := exec.Command("powershell", "-Command", psScript)
	output, err := cmd.Output()
	if err != nil {
		return "", fmt.Errorf("speech recognition error: %v", err)
	}

	text := strings.TrimSpace(string(output))
	if text == "" {
		return "", fmt.Errorf("no speech detected")
	}

	printInfo("Processing speech...")
	return text, nil
}

func printSeparator() {
	fmt.Println("\n" + strings.Repeat("-", 60) + "\n")
}

func printInfo(message string) {
	fmt.Printf("%s%s%s\n", ColorYellow, message, ColorReset)
}

func printError(message string) {
	fmt.Printf("%sError: %s%s\n", ColorRed, message, ColorReset)
}

func printUser(message string) {
	fmt.Printf("%sSir:%s %s\n", ColorGreen, ColorReset, message)
}

func printAssistant() {
	fmt.Printf("%sJarvis:%s ", ColorBlue, ColorReset)
}

func saveConversation(messages []Message) error {
	timestamp := time.Now().Format("20060102_150405")
	filename := fmt.Sprintf("conversation_%s.txt", timestamp)

	file, err := os.Create(filename)
	if err != nil {
		return err
	}
	defer file.Close()

	for _, msg := range messages {
		if msg.Role != "system" {
			fmt.Fprintf(file, "%s: %s\n\n", strings.ToUpper(msg.Role), msg.Content)
		}
	}

	printInfo(fmt.Sprintf("Conversation saved to %s", filename))
	return nil
}

func main() {
	model := "llama2-uncensored:7b"
	client := NewOllamaClient("http://localhost:11434")
	voice := NewVoiceEngine()
	voiceInputMode := false

	conversationHistory := []Message{
		{
			Role:    "system",
			Content: "You are Jarvis, a helpful AI assistant. You are speaking with your user, whom you address as \"Sir\". Please share your thoughts on any topic you find interesting however stay on track and answer the users questions directly.",
		},
	}

	fmt.Println("\n" + strings.Repeat("=", 60))
	fmt.Println("Jarvis v1.0 - Go Edition with Voice")
	fmt.Println(strings.Repeat("=", 60) + "\n")
	printInfo("Commands: exit, quit, clear, save, mute, unmute, voice, text")
	fmt.Println()

	if err := client.TestConnection(); err != nil {
		printError(err.Error())
		fmt.Println("\nPress Enter to close...")
		bufio.NewReader(os.Stdin).ReadBytes('\n')
		os.Exit(1)
	}

	voice.Speak("Jarvis online. How may I assist you, Sir?")

	reader := bufio.NewReader(os.Stdin)

	for {
		var userInput string

		if voiceInputMode {
			printInfo("Press Enter to speak (or type 'text' to switch to text input)")
			input, _ := reader.ReadString('\n')
			input = strings.TrimSpace(input)

			if strings.ToLower(input) == "text" {
				voiceInputMode = false
				printInfo("Switched to text input mode")
				continue
			} else if strings.ToLower(input) == "exit" || strings.ToLower(input) == "quit" {
				printInfo("Goodbye!")
				voice.SpeakBlocking("Goodbye, Sir.")
				return
			} else if input != "" {
				userInput = input
			} else {
				text, err := voice.Listen()
				if err != nil {
					printError(err.Error())
					continue
				}
				userInput = text
				printUser(userInput)
			}
		} else {
			fmt.Printf("%sSir:%s ", ColorGreen, ColorReset)
			input, err := reader.ReadString('\n')
			if err != nil {
				printError(fmt.Sprintf("Error reading input: %v", err))
				continue
			}
			userInput = strings.TrimSpace(input)
		}

		if userInput == "" {
			continue
		}

		switch strings.ToLower(userInput) {
		case "exit", "quit":
			printInfo("Goodbye!")
			voice.SpeakBlocking("Goodbye, Sir.")
			return

		case "clear":
			conversationHistory = []Message{conversationHistory[0]}
			printInfo("Conversation history cleared!")
			voice.Speak("Conversation history cleared.")
			continue

		case "save":
			if err := saveConversation(conversationHistory); err != nil {
				printError(fmt.Sprintf("Could not save conversation: %v", err))
			} else {
				voice.Speak("Conversation saved.")
			}
			continue

		case "mute":
			voice.ttsEnabled = false
			printInfo("Voice output muted.")
			continue

		case "unmute":
			voice.ttsEnabled = true
			printInfo("Voice output enabled.")
			voice.Speak("Voice output enabled.")
			continue

		case "voice":
			voiceInputMode = true
			printInfo("Voice input mode enabled. Press Enter to speak.")
			voice.Speak("Voice input mode enabled.")
			continue

		case "text":
			voiceInputMode = false
			printInfo("Text input mode enabled.")
			continue
		}

		conversationHistory = append(conversationHistory, Message{
			Role:    "user",
			Content: userInput,
		})

		printAssistant()

		response, err := client.Chat(model, conversationHistory)
		if err != nil {
			printError(err.Error())
			conversationHistory = conversationHistory[:len(conversationHistory)-1]
			continue
		}

		conversationHistory = append(conversationHistory, Message{
			Role:    "assistant",
			Content: response,
		})

		voice.Speak(response)

		printSeparator()
	}
}
